libs/ebpfnetsh/programs.cpp

```diff
@@ -384,31 +345,47 @@ handle_ebpf_set_program(
         return status;
     }
 
-    if (pinned.empty()) {
-        // TODO (issue #190): call ebpf_program_unpin() once it exists.
-        // The temporary API ebpf_api_unpin_object requires knowing the name a priori
-        // and we have no way to get it.
-        return ERROR_CALL_NOT_IMPLEMENTED;
-    } else {
-        // Try to find the program with the specified filename and section.
-        fd_t program_fd = _find_program_fd(filename.c_str(), section.c_str());
-        if (program_fd == ebpf_fd_invalid) {
-            std::cerr << "Program not found." << std::endl;
-            return ERROR_SUPPRESS_OUTPUT;
+    if (tags[1].bPresent) {
+        if (memcmp(&attach_type, &EBPF_ATTACH_TYPE_UNSPECIFIED, sizeof(ebpf_attach_type_t)) != 0) {
+            ebpf_result_t result = ebpf_program_attach_by_id(id, attach_type);
+            if (result != NO_ERROR) {
+                std::cerr << "error " << result << ": could not detach program" << std::endl;
+                return ERROR_SUPPRESS_OUTPUT;
+            }
+        } else {
+            int error = ebpf_program_detach_by_id(id);
+            if (error != NO_ERROR) {
+                std::cerr << "error " << error << ": could not detach program" << std::endl;
+                return ERROR_SUPPRESS_OUTPUT;
+            }
         }
+    }
 
-        // TODO (issue #83) replace ebpf_api_pin_object with ebpf_program_pin (aka bpf_program__pin)
-        // once it exists.
-        status = ebpf_object_pin(program_fd, pinned.c_str());
-        if (status != EBPF_SUCCESS) {
-            std::cerr << "error " << status << ": could not pin program" << std::endl;
-            return ERROR_SUPPRESS_OUTPUT;
-        }
+    if (tags[2].bPresent) {
+        if (pinned.empty()) {
+            // TODO (issue #190): call ebpf_program_unpin() once it exists.
+            // The temporary API ebpf_api_unpin_object requires knowing the name a priori
+            // and we have no way to get it.
```

Comment:
I think we would always require the pin path to be able to unpin the program, as a program can be pinned to multiple paths.

Reply:
Leaving that as part of issue 190 as the comment says.   We could for example try to unpin from all paths, but that's for the other issue.